### ISSUE TITLE
`.f` with multi-line output and non-null `new_col_name` results in nested tibbles

### ISSUE THREAD

**Issue by dsweber2** [cmu-delphi/epiprocess#534]
@dshemetov this is the issue I was talking about. I'm going to just route around it by not using `step_epi_slide` for now.

Compare:
```r
ress <- epi_slide(
  test_data,
  .window_size = 4,
  .f = function(slice, geo_key, ref_time_value) {
    tibble(a = mean(slice[["value"]]), b = median(slice[["value"]]))
  },
  .new_col_name = "foo"
)
ress %>% glimpse
Rows: 89
Columns: 4
Groups: geo_value [3]
$ geo_value  <chr> "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a",…
$ time_value <date> 2020-01-01, 2020-01-02, 2020-01-03, 2020-01-04, 2020-01-05…
$ value      <dbl> 0, 1, 4, 9, 16, 25, 36, 49, 64, 100, 121, 144, 169, 196, 22…
$ foo        <tibble[,2]> <tbl_df[26 x 2]>

ress
An `epi_df` object, 89 x 4 with metadata:
* geo_type  = custom
* time_type = day
* as_of     = 2020-01-31

# A tibble: 89 × 4
# Groups:   geo_value [3]
   geo_value time_value value foo$a    $b
 * <chr>     <date>     <dbl> <dbl> <dbl>
 1 a         2020-01-01     0  NA    NA  
 2 a         2020-01-02     1  NA    NA

```

with the case where `new_col_name` is `NULL`
```r
ress2 <- epi_slide(
  test_data,
  .window_size = 4,
  .f = function(slice, geo_key, ref_time_value) {
    tibble(a = mean(slice[["value"]]), b = median(slice[["value"]]))
  }
)

ress2 %>% glimpse
Rows: 89
Columns: 5
Groups: geo_value [3]
$ geo_value  <chr> "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a", "a",…
$ time_value <date> 2020-01-01, 2020-01-02, 2020-01-03, 2020-01-04, 2020-01-05…
$ value      <dbl> 0, 1, 4, 9, 16, 25, 36, 49, 64, 100, 121, 144, 169, 196, 22…
$ a          <dbl> NA, NA, NA, 3.5, 7.5, 13.5, 21.5, 31.5, 43.5, NA, NA, NA, 1…
$ b          <dbl> NA, NA, NA, 2.5, 6.5, 12.5, 20.5, 30.5, 42.5, NA, NA, NA, 1…
```
I ran into this in `step_epi_slide`, where there is always a column name added, and where nested tibbles don't interact nicely with recipes. I don't know if this is intended behavior, but I think it would probably be better to be consistent about always appending normal-ish row data. in the case of a name being specified, I was expecting `"fooa"` and `"foob"`, or maybe `"foo_a"` and `"foo_b"`.

Also, the docs are thankfully pretty precise about the expected input of `.f`. They are not, however, very precise about the expected output.

**Comment by dshemetov** [comment]
Does having to do `foo$column` to select columns downstream pose problems for epipredict workflows in a way that `foo_column` doesn't or is it just unexpected?

As for the docs issue: this is mentioned in the Details section of the slide docs [here](https://cmu-delphi.github.io/epiprocess/reference/epi_slide.html).

> , fn_returning_a_data_frame(.x) will unpack the output of the function into multiple columns in the result.
>
> Named expressions evaluating to data frames will be placed into [tidyr::pack](https://tidyr.tidyverse.org/reference/pack.html)ed columns.

Though admittedly this information is both:
- buried in the already-extremely dense epi_slide documentation 
- is quite terse and probably doesn't make sense to people who aren't familiar with packing and unpacking (which was me until about a month ago when I dove into the slide changes)

We've been in maintenance mode wrt slide docs, just making sure they aren't incorrect, through the recent migrations, but we really need to take a pass through it for readability for people who are not already experts with `tidyr::pack` and other tidyselect features.

**Comment by dsweber2** [comment]
yes, they came up with a pretty way to print it, but if you're e.g. listing the column names, you only get the top level, and not `foo$column`
```r
> ex <- tibble(a = c(1,2), b = tibble(c=c(1,3),d=c(1,5)))
> ex
# A tibble: 2 × 2
      a   b$c    $d
  <dbl> <dbl> <dbl>
1     1     1     1
2     2     3     5
> names(ex)
[1] "a" "b"
> ex %>% select(ends_with("c"))
# A tibble: 2 × 0

```
As far as most functions are concerned, there are only 2 columns. Since recipes relies on column logic quite a bit, it can't handle nested tibbles.

**Comment by dsweber2** [comment]
This issue is relatively low priority because it's possible to route around by just naming things yourself, either in the function or using `rename`, but we should probably avoid the nesting eventually.

**Comment by dshemetov** [comment]
Ah I see, so it creates a nested structure which doesn't seem to be well-supported by tidyselect. @brookslogan do you know of a workaround for this? I think the intention of removing `names_sep` and nesting like this was to avoid name collisions.

**Comment by brookslogan** [comment]
-- Context --

This behavior was intentional.  My goal was to
- avoid the name prefix spam situations people got into with `epix_slide` for pseudoprospective forecasting; removing prefixing altogether seemed like one way
  - (& avoid some column name conflicts encountered when trying to use a simple workaround to the prefix spam, manually unnesting)
- make `out = f(.x)` consistent with `~ f(.x), .new_col_name = "out"`, and consistent with tidyverse; this meant packed columns; try `tibble(x = 1, y = tibble(x2 = x + x))`

Tidyeval here is following data-masking (used by `mutate` etc.), not tidyselect (used by `epi_slide_opt`, `across`).

-- My guess at tidyverse context --

> yes, they came up with a pretty way to print it, but nested tibbles are not even dplyr-friendly. If you're e.g. listing the column names, you only get the top level, and not foo$column

I'm pretty sure this is intended behavior.  It's meant to hide/bundle those columns.

> As far as most functions are concerned, there are only 2 columns. Since recipes relies on column logic quite a bit, it can't handle nested tibbles.

Packed tibbles seem like they were meant to be a solution to a lot of such naming issues in tidyverse, though maybe recipes doesn't use this solution (and allowing for multiple roles is more flexible than bundling).  Instead of relying on naming schemes & patterns to maybe select the intended set of columns, you can pack all the relevant columns into one bundle when you generate them.  Or you can pack a column or columns with variable names into a column with a fixed name, rather than using/struggling with `!! :=`.

Tidyverse's solution to suffixes is by using `across`.

-- Idea --

- We could add prefixing/suffixing features onto slides as an alternative to `.new_col_name`.  Suffixes really seem to be what we want for `epi_slide_opt`.
- We could try to support `across` more natively in slides; right now you'd have to do something like `.f = ~ .x %>% dplyr::reframe(across(your_col_names, list(your_func_name = your_func)))}`
